fix: suggest hidden:true with getByRole if element is not accessible

marcosvega91 opened the pull request:
<!-- What changes are being made? (What feature/bug is being fixed here?) -->

**What**: If an element is inaccessible `getSuggestedQuery` should suggest to use `hidden:true` with `getByRole` 

<!-- Why are these changes necessary? -->

**Why**: Answering to this issue https://github.com/testing-library/dom-testing-library/issues/744, I wanted to share a link of `testing-playground` but the query was not suggested as I expected.  

<!-- How were these changes implemented? -->

**How**: If the element is inaccessible and the query is `getByRole` I have added to `queryArgs` an `hidden:true`

<!-- Have you done all of these things?  -->

**Checklist**:

<!-- add "N/A" to the end of each line that's irrelevant to your changes -->

<!-- to check an item, place an "x" in the box like so: "- [x] Documentation" -->

- [ ] Documentation added to the
      [docs site](https://github.com/testing-library/testing-library-docs)
- [X] Tests
- [ ] Typescript definitions updated
- [X] Ready to be merged
      <!-- In your opinion, is this ready to be merged as soon as it's reviewed? -->

<!-- feel free to add additional comments -->

eps1lon reviewed (changes requested):
It shouldn't just suggest it but rather say that the/a element is not part of the a11y tree. 

The element being removed from the a11y tree has strong implications for users relying on screen readers. It's equivalent to the element being invisible for a sighted user.

I fear people will just randomly add aria-hidden and query for it if we explicitly recommend it.

marcosvega91 commented:
Yes it was something to discuss. I think that we should add this to testing playground too. Maybe we should do a console.warn? While on testing playground show a text ?

marcosvega91 commented:
what do you think of this ?
I have added a `warning` prop with a message, maybe this will be useful for other things. On testing-playground side we can show the warning message

p.s. I have added some types to `suggestions` 

```diff
diff --git a/src/suggestions.js b/src/suggestions.js
index f584e7b..0d23434 100644
--- a/src/suggestions.js
+++ b/src/suggestions.js
@@ -39,6 +39,8 @@ function getRegExpMatcher(string) {
 }
 
 function makeSuggestion(queryName, element, content, {variant, name}) {
+  let warning = ''
+
   const queryArgs = [
     queryName === 'Role' || queryName === 'TestId'
       ? content
@@ -51,6 +53,10 @@ function makeSuggestion(queryName, element, content, {variant, name}) {
 
   if (queryName === 'Role' && isInaccessible(element)) {
     queryArgs.push({hidden: true})
+    warning = `Element is inaccessible. This means that the element and all his children are invisible to screen readers.
+    If you are using aria-hidden prop make sure this is the right choice for your case.
+    `
+    console.warn(warning)
   }
 
   const queryMethod = `${variant}By${queryName}`
@@ -60,6 +66,7 @@ function makeSuggestion(queryName, element, content, {variant, name}) {
     queryMethod,
     queryArgs,
     variant,
+    warning,
     toString() {
       let [text, options] = queryArgs
 
diff --git a/types/suggestions.d.ts b/types/suggestions.d.ts
index 17e48d6..8c1e68c 100644
--- a/types/suggestions.d.ts
+++ b/types/suggestions.d.ts
@@ -1,5 +1,13 @@
+export type QueryArg = {
+  [key: string]: RegExp | boolean
+}
+
 export interface Suggestion {
   queryName: string
+  queryMethod: string
+  queryArgs: QueryArg[]
+  variant: string
+  warning: string
   toString(): string
 }
```

nullromo commented:
```
-    warning = `Element is inaccessible. This means that the element and all his children are invisible to screen readers.
-    If you are using aria-hidden prop make sure this is the right choice for your case.

+    warning = `Element is inaccessible. This means that the element and all its children are invisible to screen readers.
+    If you are using the aria-hidden prop, make sure this is the right choice for your case.
```

eps1lon commented:
> what do you think of this ? [...]

-- https://github.com/testing-library/dom-testing-library/pull/745#issuecomment-676369200

:+1:  That is way clearer. There's always some risk involved but I hope this won't cause too much damage.

marcosvega91 commented:
I have added the message as we discussed above 👍
